### PR TITLE
Replace XSLT with TS

### PR DESCRIPTION
**Pull request by nleanba**
Open todos:

- [x] figures
- [ ] refactor code to be more sensible (currently, a lot is a 1:1 translation from XSLT and thus very convoluted. If it was immediately obvious to me how to streamline somthing, it was done, but XSLT is hard)
- [x] integration into pipeline
- [x] running a test run over all treatments and comparing for (unwanted) changes

### CODE REVIEW

**Comment by nleanba** [comment]
manually checked all treatments until 006E64249F0BFFC1A6E1FAC0364EF948 (going alphabetically by ID)

they are all fine*

A notable change is that it changes a bunch of `treatsTaxonName` to `definesTaxonName`, as far as i can tell this seems correct

some changes that I consider irrelevant or even improvements:
- it no longer doubly-escapes special characters in authorities in urls
- some taxon-name and taxon-concept uris have changed
- it correctly puts the kingdom for trt:treatsTaxonName
- some taxon-name/taxon-concept uris are no longer all-uppercase for no apparent reason
- some uris are more consistent (notably, the publication is now always linked correctly)
- it adds a lot more information about taxonNames and taxonConcepts mentioned, i.e. it expands the details for all tn/tc that are mentioned
- it adds some materialsCitations that were ignored previously (due to missing specimen code)
-

**Comment by nleanba** [comment]
Weird change in 0129163AB112602AFCCAFBD9FE4DF90D:
the publication uri is changed to the one provided as `docSource`, which seems not to actually point to the article itself (unlike the uri generated by XSLT which is from `ID-DOI`). Is this a bug in the XML or in the XSLT-replacement?

**Comment by nleanba** [comment]
Checked until 016512208D171293F4B2C381B36B1F22

All fine notwithstanding changes mentioned above

I think it can replace the xslt like this, fixing potential other errors when they are noticed.

@retog opinions?

File: src/gg2rdf.ts [hunk 1]
```diff
@@ -0,0 +1,1208 @@
...
+*/
+
+import { DOMParser, parseArgs } from "./deps.ts";
+import { Element } from "https://esm.sh/v135/linkedom@0.16.8/types/interface/element.d.ts";
+
+class Subject {
+  uri: string;
+  properties: { [key: string]: Set<string> };
+
+  constructor(uri: string) {
```
Comment (retog) [inline src/gg2rdf.ts]: The modifier `public` on the parameter would obsolete the assignment and field

File: src/gg2rdf.ts [hunk 2]
```diff
@@ -0,0 +1,1208 @@
...
+  /** sorted with comments first, and type last */
+  get propNames() {
+    return Object.getOwnPropertyNames(this.properties).sort((a, b) => {
+      if (a === b) return 0;
+      if (
+        a.startsWith("#") && b.startsWith("#")
+      ) return a.slice(1) < b.slice(1) ? -1 : 1;
+      if (a.startsWith("#")) return -1;
+      if (b.startsWith("#")) return 1;
+      if (a === "a") return 1;
```
Comment (retog) [inline src/gg2rdf.ts:18]: What's this special case for?

File: src/gg2rdf.ts [hunk 3]
```diff
@@ -0,0 +1,1208 @@
...
+
+  if (!flags.input) throw new Error("No input file provided");
+  if (!flags.output) flags.output = flags.input + ".ttl";
+  console.log("aha");
+  gg2rdf(flags.input, flags.output);
+}
+
+export function gg2rdf(inputPath: string, outputPath: string) {
+  const document = new DOMParser().parseFromString(
+    Deno.readTextFileSync(inputPath).replaceAll(/(<\/?)mods:/g, "$1MODS"),
```
Comment (retog) [inline src/gg2rdf.ts:46]: Could you explain the replace?

File: src/gg2rdf.ts [hunk 4]
```diff
@@ -0,0 +1,1208 @@
...
+        console.warn(
+          "Warning: treatment taxon is missing ancestor kingdom, defaulting to 'Animalia'",
+        );
+        output(
+          "# Warning: treatment taxon is missing ancestor kingdom, defaulting to 'Animalia'",
+        );
+      }
+    }
+    if (errors.length) {
+      throw new Error(
```
Comment (retog) [inline src/gg2rdf.ts:183]: RDF/XML?

**Review by retog — COMMENTED** [review]
A few comments and questions. Overall looking good but I'll look at it in more detail.

File: src/gg2rdf.ts (hunk 2 above)
Comment (nleanba) [reply src/gg2rdf.ts:18]: for the type, always moving the type to the end.

the other part is to sort comments (properties starting with `#`) to the front, to ensure valid output (the last property must not be a comment)

File: src/gg2rdf.ts (hunk 3 above)
Comment (nleanba) [reply src/gg2rdf.ts:46]: beacuse i could not figure out how to write `querySelector`s for namespaced elements, so i just rename all `mods`-namespace elements

File: src/gg2rdf.ts (hunk 4 above)
Comment (nleanba) [reply src/gg2rdf.ts:183]: error message is copy-pasted from xslt

you are correct that that makes no sense anymore, I will change it

**Comment by retog** [comment]
Why is `outputProperties`only used in `makePublication`?

**Comment by nleanba** [comment]
> Why is `outputProperties`only used in `makePublication`?

`outputProperties` was the first way ive done it, everything else has been updated to use `Subject`, but I didn't think changing it for `makePublication` would be worth it.

**Comment by retog** [comment]
I've added `fish` to the container but I still can't execute the script

```
root@fbc1a93c34b7:/workspaces/gg2rdf# ./test_noxslt.fish ./ex.xml 
File ./ex.ttl doesn't exist! Aborting
root@fbc1a93c34b7:/workspaces/gg2rdf# ls ex.xml 
ex.xml
root@fbc1a93c34b7:/workspaces/gg2rdf# ./test_noxslt.fish `pwd`/ex.xml 
File /workspaces/gg2rdf/ex.ttl doesn't exist! Aborting
```

**Comment by retog** [comment]
> I didn't think changing it for `makePublication` would be worth it.

I tend to disagree. It makes the NOTES at the top of the file inaccurate, and even if that is fixed it still increases complexity. I think the additional time needed when making future changes far outweighs the tedious but relatively small amount of work to make it consistent now.

**Comment by nleanba** [comment]
> I've added `fish` to the container but I still can't execute the script
> 
> ```
> root@fbc1a93c34b7:/workspaces/gg2rdf# ./test_noxslt.fish ./ex.xml 
> File ./ex.ttl doesn't exist! Aborting
> root@fbc1a93c34b7:/workspaces/gg2rdf# ls ex.xml 
> ex.xml
> root@fbc1a93c34b7:/workspaces/gg2rdf# ./test_noxslt.fish `pwd`/ex.xml 
> File /workspaces/gg2rdf/ex.ttl doesn't exist! Aborting
> ```

The error is because it checks for {$ttlReferenceDir}/ex.**ttl**, which probaly does not exist in the docker container.

`./test_noxslt.fish` was never intended to be run in the container, but is only for manual testing. If you wish to use it in a container, you should already be running an interactive shell in the container, install fish from there.

If you wish to simply check what output `gg2rdf.ts` produces, just call `deno run --allow-read --allow-write ./gg2rdf.ts -i <xml filename> -o <ttl filename>`

File: Dockerfile [hunk 5]
```diff
@@ -1,8 +1,10 @@
 FROM denoland/deno:ubuntu-1.39.0
 
-# Install cron
 RUN apt update
-RUN DEBIAN_FRONTEND=noninteractive apt install -y raptor2-utils openjdk-11-jdk git curl
+RUN DEBIAN_FRONTEND=noninteractive apt install -y software-properties-common
+RUN DEBIAN_FRONTEND=noninteractive apt-add-repository -y ppa:fish-shell/release-3 
+RUN apt update
```
Comment (nleanba) [inline Dockerfile]: why a second apt update?

File: Dockerfile
Comment (nleanba) [inline Dockerfile:1]: there is no reason imo to add fish to the container

**Comment by nleanba** [comment]
> > I didn't think changing it for `makePublication` would be worth it.
> 
> I tend to disagree. It makes the NOTES at the top of the file inaccurate, and even if that is fixed it still increases complexity. I think the additional time needed when making future changes far outweighs the tedious but relatively small amount of work to make it consistent now.

i have changed this now

File: Dockerfile
Comment (nleanba) [reply Dockerfile:1]: i have removed it again, so that the container actually builds again

**Comment by retog** [comment]
Weird that it didn't! I committed from the working dev container.

**Comment by nleanba** [comment]
@retog can we merge this?

**Comment by retog** [comment]
Experimenting with 000040332F2853C295734E7BD4190F05. I see the current ttl version has 106 triples,the one generated by the new transformer 118.

These are the additional triples:
```
 <http://taxon-concept.plazi.org/id/Animalia/Saigona_Matsumura_1910> <http://plazi.org/vocab/treatment#hasTaxonName> <http://taxon-name.plazi.org/id/Animalia/Saigona> .
 <http://taxon-concept.plazi.org/id/Animalia/Saigona_Matsumura_1910> <http://rs.tdwg.org/dwc/terms/class> "Insecta" .
 <http://taxon-concept.plazi.org/id/Animalia/Saigona_Matsumura_1910> <http://rs.tdwg.org/dwc/terms/family> "Dictyopharidae" .
 <http://taxon-concept.plazi.org/id/Animalia/Saigona_Matsumura_1910> <http://rs.tdwg.org/dwc/terms/genus> "Saigona" .
 <http://taxon-concept.plazi.org/id/Animalia/Saigona_Matsumura_1910> <http://rs.tdwg.org/dwc/terms/kingdom> "Animalia" .
 <http://taxon-concept.plazi.org/id/Animalia/Saigona_Matsumura_1910> <http://rs.tdwg.org/dwc/terms/order> "Hemiptera" .
 <http://taxon-concept.plazi.org/id/Animalia/Saigona_Matsumura_1910> <http://rs.tdwg.org/dwc/terms/phylum> "Arthropoda" .
 <http://taxon-concept.plazi.org/id/Animalia/Saigona_Matsumura_1910> <http://rs.tdwg.org/dwc/terms/rank> "genus" .
 <http://taxon-concept.plazi.org/id/Animalia/Saigona_Matsumura_1910> <http://rs.tdwg.org/dwc/terms/scientificNameAuthorship> "Matsumura, 1910" .
 <http://taxon-concept.plazi.org/id/Animalia/Saigona_Matsumura_1910> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://filteredpush.org/ontologies/oa/dwcFP#TaxonConcept> .
43a54
 <http://taxon-concept.plazi.org/id/Animalia/Saigona_baiseensis_Zheng_2021> <http://rs.tdwg.org/dwc/terms/scientificNameAuthorship> "Zheng & Chen, 2021" .
100a112
 <http://treatment.plazi.org/id/000040332F2853C295734E7BD4190F05> <http://purl.org/dc/elements/1.1/title> "Saigona baiseensis Zheng & Chen 2021, sp. nov." .
```
The additional triples look sound, the previously missing title matches the one on https://treatment.plazi.org/id/000040332F2853C295734E7BD4190F05

I do see two error messages:
```
Error: Invalid Authority for <http://taxon-concept.plazi.org/id/Animalia/fulgoroidesINVALID>
Error: Invalid Authority for <http://taxon-concept.plazi.org/id/Animalia/fulgoroidesINVALID>
```

**Comment by retog** [comment]
Regarding
```
`output(...)` should not be assumed to run synchronous,
  and all data passed to it should still be valid under reordering of calls.
```
In turtle the order plays a role with regard to base and prefix, so I think this requirement should be dropped.

**Comment by nleanba** [comment]
> I do see two error messages:
> 
> ```
> Error: Invalid Authority for <http://taxon-concept.plazi.org/id/Animalia/fulgoroidesINVALID>
> Error: Invalid Authority for <http://taxon-concept.plazi.org/id/Animalia/fulgoroidesINVALID>
> ```

Those Errors indicate that it found mentions of these taxa, but it could not figure out their authorities to turn them into citations (augments or deprecates). The rdf output matches the behaviour of the xslt, but with more insight into why it was generated in the way it was.

File: example-data/000040332F2853C295734E7BD4190F05.xml
Comment (nleanba) [inline example-data/000040332F2853C295734E7BD4190F05.xml:1]: I don't really see the point of having this?
I'd rather test as directly on a bunch files from a current clone of the treatment-xml repo, given that no single file can cover all edge cases and the xml may change

Comment (retog) [reply example-data/000040332F2853C295734E7BD4190F05.xml:1]: I think it greatly facilitates debugging

**Comment by nleanba** [comment]
@retog did you rebase this into main?

this is rather inelegant, as now all my commits are marked as unverified.

i think a merge commit would have been better

**Comment by retog** [comment]
> @retog did you rebase this into main?
> 
> this is rather inelegant, as now all my commits are marked as unverified.
> 
> i think a merge commit would have been better

I did. I didn't know about this implication.
